README.md
# Shoket Client for PHP

[![Latest Version on Packagist](https://img.shields.io/packagist/v/shoket/php-shoket.svg?style=flat-square)](https://packagist.org/packages/shoket/php-shoket)
[![Tests](https://github.com/alphaolomi/php-shoket/actions/workflows/run-tests.yml/badge.svg?branch=main)](https://github.com/alphaolomi/php-shoket/actions/workflows/run-tests.yml)
[![Total Downloads](https://img.shields.io/packagist/dt/shoket/php-shoket.svg?style=flat-square)](https://packagist.org/packages/shoket/php-shoket)


A Truly simple payments for your business. Shoket makes online payments more easier for all kind of business in Tanzania. From startups, small businesses to medium and large businesses.

## Documentation

You can find the documentation for this package on [Docs](https://alphaolomi.github.io/php-shoket/)

## Installation

> Requires PHP 8.0 or higher

> For Laravel users there is a dedicated intergration package [laravel-shoket](https://github.com/alphaolomi/laravel-shoket)

You can install the package via composer:

```bash
composer require shoket/php-shoket
```

## Usage

```php
// Create a new instance of the client
$shoket = new Shoket(['apiSecret' => 'your-api-key']);

// Make a payment request
$response = $shoket->makePaymentRequest([
    "amount" => "5000",
    "customer_name" => "John Doe",
    "email" => "john@user.com",
    "number_used" => "255612345678",
    "channel" => "Tigo",
]);

// Print the response
print_r($response);
```

## API Available

-   makePaymentRequest
-   verifyPayment


##  Make payment request API
Example
```php
// Create a new instance of the client
$shoket = new Shoket(['apiSecret' => 'your-api-key']);

// Make a payment request
$response = $shoket->makePaymentRequest([
    "amount" => "5000",
    "customer_name" => "John Doe",
    "email" => "john@user.com",
    "number_used" => "255612345678",
    "channel" => "Tigo",
]);

// Print the response
var_dump($response);
```

##   verifyPayment API
Example
```php

$shoketClient = new Shoket(['apiSecret' => 'your-api-key']);

// Get the payment reference from the a successful payment request
// Sample: adz49dS428b7kbDTdG4MN
$reference = 'your-reference-number'; 

$response  = $shoketClient->verifyPaymentRequest($reference,[
    "provider_name"=> "Vodacom",
    "provider_code"=> "MPESA"
]);

var_dump($response);
```
## Testing
Uses [Pest Testing Framework](https://pestphp.com/)

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

-   [Alpha Olomi](https://github.com/alphaolomi)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `php-shoket` will be documented in this file.

## v0.0.5 - 2022-06-24

### Updates

- Update version constrain with php-shoket

<br>
**Full Changelog**: https://github.com/alphaolomi/php-shoket/compare/v0.0.3...v0.0.5

## v0.0.3 - 2022-05-25

## What's Changed

- fix: typo in verify method 730dcbaf34691870a50b66a2f386730bf3b8b669
- build(deps): bump dependabot/fetch-metadata from 1.2.1 to 1.3.1 by @dependabot in https://github.com/alphaolomi/php-shoket/pull/15

**Full Changelog**: https://github.com/alphaolomi/php-shoket/compare/v0.0.2...v0.0.3

## v0.0.2 - 2022-02-27

- Fix minor typo in URL endpoint

## v0.0.1 - 2022-02-27

## What's Changed

- build(deps): bump dependabot/fetch-metadata from 1.1.1 to 1.2.0 by @dependabot in https://github.com/alphaolomi/php-shoket/pull/5
- feat: 🎸 add apis by @alphaolomi in https://github.com/alphaolomi/php-shoket/pull/6

## New Contributors

- @dependabot made their first contribution in https://github.com/alphaolomi/php-shoket/pull/5
- @alphaolomi made their first contribution in https://github.com/alphaolomi/php-shoket/pull/6

**Full Changelog**: https://github.com/alphaolomi/php-shoket/commits/v0.0.1

## 0.0.1 - 2022-02-27

## What's new

- initial release

## Features

- makePaymentRequest
- verifyPaymentRequest


tests/ShoketTest.php
<?php

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use Shoket\SDK\Shoket;

it('can test', function () {
    expect(true)->toBeTrue();
});


it('can instantiate without Guzzle/Http Client', function () {
    $s = new Shoket(['apiSecret' => '123456789',]);

    expect($s)->toBeInstanceOf(Shoket::class);
});

it('can change Api secret', function () {
    $s = new Shoket(['apiSecret' => '123456789',]);
    $apiSecret = '987654321';
    $s->setApiKey($apiSecret);

    expect($s)->toBeInstanceOf(Shoket::class);
    expect($s->getApiKey())->toBe($apiSecret);
});

it('throws InvalidArgumentException if API keys  is missing', function () {
    new Shoket();
})->throws(InvalidArgumentException::class);

it('can instantiate with Custom Guzzle/Http Client', function () {
    $mock = new MockHandler([new Response(200, ['X-Foo' => 'Bar'], 'Hello World')]);
    $handlerStack = HandlerStack::create($mock);
    $client = new Client(['handler' => $handlerStack]);

    $s = new Shoket([
        'apiSecret' => '123456789',
    ], $client);

    expect($s)->toBeInstanceOf(Shoket::class);
});


it('can make payment request', function () {
    $mock = new MockHandler([new Response(200, ['X-Foo' => 'Bar'], '{
        "Status": "success",
        "customer": {
          "customer_name": "376FcD3gbidW",
          "email": "user@user.com",
          "id": 64043
        },
        "data": {
          "amount": 64000,
          "channel": "Halotel",
          "currency": "TSH",
          "number_used": "0663803078",
          "status": "Success",
          "transaction_date": "2021-06-27 15:08:59.917691"
        },
        "message": "Transaction is completed",
        "reference": "adz49dS428b7kbDTdG4MN"
      }')]);
    $handlerStack = HandlerStack::create($mock);
    $client = new Client(['handler' => $handlerStack]);

    $s = new Shoket([
        'apiSecret' => '123456789',
    ], $client);

    try {
        $res = $s->makePaymentRequest([
            "amount" => "5000",
            "customer_name" => "John Doe",
            "email" => "john@user.com",
            "number_used" => "255612345678",
            "channel" => "Tigo",
        ]);
        expect($res)->toBeArray();
    } catch (\GuzzleHttp\Exception\GuzzleException $e) {
    }
});


it('can make verify payment requests', function () {
    $mock = new MockHandler([new Response(200, ['X-Foo' => 'Bar'], '{
        "Status": "Ok",
        "customer": {
          "email": "user@user.com",
          "customer_name": "John",
          "id": 64043
        },
        "data": {
          "amount": 64000,
          "channel": "Halotel",
          "currency": "TSH",
          "number_used": "0663803078",
          "status": "Success",
          "transaction_date": "2021-06-27 15:08:59.917691"
        },
        "message": "Transaction is completed",
        "reference": "adz49dS428b7kbDTdG4MN"
      }
      ')]);
    $handlerStack = HandlerStack::create($mock);
    $client = new Client(['handler' => $handlerStack]);

    $s = new Shoket([
        'apiSecret' => '123456789',
    ], $client);

    try {
        $res = $s->verifyPaymentRequest('OB3J177Lqnp6Rg6wHqr3q', [
            "provider_name" => "Vodacom",
            "provider_code" => "MPESA",
        ]);
        expect($res)->toBeArray();
    } catch (\GuzzleHttp\Exception\GuzzleException $exception) {
    }
});


src/Shoket.php
<?php

namespace Shoket\SDK;

use GuzzleHttp\Client;

/**
 * Shoket SDK
 * @package Shoket\SDK
 * @example
 * ```php
 * $shoket = new Shoket(['apiSecret' => 'your-api-key']);
 * ```
 */
class Shoket
{
    /** @var \GuzzleHttp\Client Client */
    private Client $client;

    /** @var string apiVersion */
    private string $apiVersion = 'v1';

    /** @var string apiSecret */
    private string $apiKey;

    public function __construct(array $options = [], Client $client = null)
    {
        if (array_key_exists('apiSecret', $options)) {
            $this->apiKey = $options['apiSecret'];
        } else {
            throw new \InvalidArgumentException('API Secret is required, To obtain visit https://developers.shoket.co/');
        }

        $this->client = $client instanceof Client ? $client : new Client([
            'base_uri' => 'https://api.shoket.co/' . $this->apiVersion . '/',
            'headers' => [
                'Accept' => 'application/json',
                'Authorization' => 'Bearer ' . $this->apiKey,
                'Content-Type' => 'application/json',
            ],
        ]);
    }

    /**
     * Make payment request
     * @param array $data
     * @return mixed
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     * @example
     * ```php
     * //...
     * $shoket->makePayment([
     *  "amount" => "5000",
     *  "customer_name" => "John Doe",
     *  "email" =>  "john@user.com",
     *  "number_used" => "255612345678",
     *  "channel" =>  "Tigo"
     * ]);
     * ```
     *
     */
    public function makePaymentRequest(array $data): mixed
    {
        $response = $this->client->request('POST', 'charge/', ['json' => $data]);

        return json_decode($response->getBody()->getContents(), true);
    }

    /**
     * Verify Payment request.
     *
     * @param string $reference
     * @param array $data
     * @return mixed
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     * @example
     * ```php
     * $shoketClient->verifyPaymentRequest('123456789',[
     *  "provider_name"=> "Vodacom",
     *  "provider_code"=> "MPESA"
     * ]);
     * ```
     *
     */
    public function verifyPaymentRequest(string $reference, array $data): mixed
    {
        $response = $this->client->request('POST', 'verify/' . $reference, ['json' => $data]);

        return json_decode($response->getBody()->getContents(), true);
    }

    /**
     * @return Client
     */
    public function getClient(): Client
    {
        return $this->client;
    }

    /**
     * @param Client $client
     */
    public function setClient(Client $client): void
    {
        $this->client = $client;
    }

    /**
     * @return string
     */
    public function getApiKey(): string
    {
        return $this->apiKey;
    }

    /**
     * @param string $apiKey
     */
    public function setApiKey(string $apiKey): void
    {
        $this->apiKey = $apiKey;
    }
}
